Matomo application logs can now be written in syslog and errorlog

mwithheld opened the pull request:
### Description:

Issue #9400 Add log handlers syslog and errorlog; Add config for syslog ident; Document options.

### Review

* [ ] Functional review done
* [ ] Potential edge cases thought about (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] Code review done
* [ ] Tests were added if useful/possible
* [ ] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

mwithheld committed:
Add log handlers syslog and errorlog

Issue#9400

mwithheld committed:
Merge branch 'matomo-org:4.x-dev' into patch-2

mwithheld committed:
Add log handlers syslog and errorlog

mwithheld committed:
syslog/errorlog: fix namespace; add log.syslog.ident

mwithheld committed:
syslog/errorlog: Document options

mwithheld commented:
This is an update to PR #17745

diosmosis commented on plugins/Monolog/config/config.php:
```diff
@@ -97,6 +99,13 @@
         ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
+    
+    '\Monolog\Handler\ErrorLogHandler' => DI\create()
+        ->constructor(null, DI\get('log.level.errorlog'))
```
Looking at the ErrorHandler constructor, it seems like this shouldn't be set to null. If you want to only set the level parameter, and use defaults for the rest, you can use:

```
->constructorParameter('level', DI\get('log.level.errorlog'))
```

mwithheld replied:
That is one nifty dependency injection system you've got going!  So would it be like this?
```
   '\Monolog\Handler\ErrorLogHandler' => DI\create()
   ->constructorParameter('level', DI\get('log.level.errorlog'))
   ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
```

Thanks so much for the guidance dizzy!

diosmosis replied:
@mwithheld yes, that looks correct! And yes it is definitely a nifty DI system! The library is called php-di and the creator, @/mnapoli, worked w/ us for a while so our integration works rather well.

diosmosis commented:
@mwithheld Thanks for fixing the issues! I left one other comment. After it's addressed it should be good to merge.

mwithheld committed:
ErrorLogHandler constructor: Avoid null

Avoid null in ErrorLogHandler constructor call: Use level param; default others

mwithheld committed:
Merge branch 'matomo-org:4.x-dev' into patch-2

mwithheld commented:
Added commit 14992a8 ErrorLogHandler constructor: Avoid null

diosmosis committed:
missing comma

diosmosis commented on plugins/Monolog/config/config.php:
```diff
@@ -97,6 +99,13 @@
+        ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
+
+    '\Monolog\Handler\SyslogHandler' => DI\create()
+        ->constructor(DI\get('log.syslog.ident'), 'syslog', DI\get('log.level.syslog')),
```
Apologies for the back and forth, but one last question here: was it intentional not to use `->method('setFormatter', DI\get('log.lineMessageFormatter.file'))` here?

mwithheld commented:
Good catch - the discussion is welcome :)  That was not intended - I'm just
not used to using GitHub for editing PRs.

On Thu, Jul 15, 2021 at 7:20 PM dizzy ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In plugins/Monolog/config/config.php
> <https://github.com/matomo-org/matomo/pull/17764#discussion_r670919393>:
>
> > @@ -97,6 +99,13 @@
>      'Piwik\Plugins\Monolog\Handler\FileHandler' => DI\create()
>          ->constructor(DI\get('log.file.filename'), DI\get('log.level.file'))
>          ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
> +
> +    '\Monolog\Handler\ErrorLogHandler' => DI\create()
> +        ->constructorParameter('level', DI\get('log.level.errorlog'))
> +        ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
> +
> +    '\Monolog\Handler\SyslogHandler' => DI\create()
> +        ->constructor(DI\get('log.syslog.ident'), 'syslog', DI\get('log.level.syslog')),
>
> Apologies for the back and forth, but one last question here: was it
> intentional not to use ->method('setFormatter',
> DI\get('log.lineMessageFormatter.file')) here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matomo-org/matomo/pull/17764#pullrequestreview-707953192>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGCSOXPWKB6RSVBKTMBCRDTX6JQRANCNFSM5AKBNYHA>
> .
>

diosmosis committed:
fix two di definitions

diosmosis committed:
Merge branch '4.x-dev' into patch-2

diosmosis commented:
@mwithheld I've merged the PR, thanks for going through the trouble of creating the PR and following up on it! The contribution is greatly appreciated.

mattab commented:
@diosmosis could you please document this new capability in https://matomo.org/faq/troubleshooting/faq_115/ ? Thanks

diosmosis commented:
documented

orlitzky commented on plugins/Monolog/config/config.php:
```diff
@@ -97,6 +99,14 @@
+        ->method('setFormatter', DI\get('log.lineMessageFormatter.file')),
+
+    '\Monolog\Handler\SyslogHandler' => DI\create()
+        ->constructor(DI\get('log.syslog.ident'), 'syslog', DI\get('log.level.syslog'))
```
Is this logging to the "syslog" facility? I would guess that "user" is most appropriate, but "syslog" is definitely not correct =)

(Discovered while trying to match on the facility and program name using syslog-ng.)

Lazy reference: https://en.wikipedia.org/wiki/Syslog#Facility

mwithheld replied:
You are correct this should be 'user'.
Should probably use 
```
->constructorParameter('ident'=>DI\get('log.syslog.ident'))
->constructorParameter('level'=>DI\get('log.level.syslog'))
```

diosmosis commented:
Hi @mwithheld, thanks for creating the change so quickly! Would you be able to put it in a new pull request? This one has already been merged so it's no longer useful to push here.

orlitzky commented:
Ok, thanks! And thanks for adding syslog support in the first place. Obviously, I was waiting for it =)
